src/Servers/Kestrel/Transport.Sockets/src/Internal/IOQueue.cs

```diff
@@ -22,35 +21,66 @@ public override void Schedule(Action<object> action, object state)
                 State = state
             };
 
+            // Order is important here with Execute().
+
+            // 1. Enqueue prior to checking _doingWork.
             _workItems.Enqueue(work);
 
-            lock (_workSync)
+            // 2. MemoryBarrier to ensure ordering of Write (workItems.Enqueue) -> Read (_doingWork) is preserved,
+            // as order is reversed between Schedule and Execute, and they are two different memory locations.
+            Thread.MemoryBarrier();
+
+            // 3. Fast check if already doing work, don't need Volatile here due to explicit MemoryBarrier above
+            if (_doingWork == 0)
             {
-                if (!_doingWork)
+                // 4. Not working, set as working, and check if it was already working (via atomic Interlocked).
+                var submitWork = Interlocked.Exchange(ref _doingWork, 1) == 0;
+
+                if (submitWork)
                 {
+                    // 5. Wasn't working, schedule.
                     System.Threading.ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
-                    _doingWork = true;
                 }
             }
         }
 
         void IThreadPoolWorkItem.Execute()
         {
+            var workItems = _workItems;
             while (true)
             {
-                while (_workItems.TryDequeue(out Work item))
+                while (workItems.TryDequeue(out Work item))
                 {
                     item.Callback(item.State);
                 }
 
-                lock (_workSync)
+                // Order is important here with Schedule(...).
+
+                // 1. Set _doingWork (0 == false) prior to checking .IsEmpty
+                // Don't need Volatile here due to explicit MemoryBarrier below
+                _doingWork = 0;
```

Comment:
Between here and the Interlocked.Exchange, it is possible for you to have multiple threads executing the code in this region.  I assume that's ok?  I can't think of anything that could go wrong as a result, but wanted to highlight it.